source/fe/mapping_q1.cc

```diff
@@ -92,6 +93,246 @@ namespace internal
+      const double xi_denominator0 = eta*x3 - x1*(eta - 1) + subexpr0;
+      const double max_x = std::max(std::max(std::abs(x0), std::abs(x1)),
+                                    std::max(std::abs(x2), std::abs(x3)));
+
+      if (std::abs(xi_denominator0) > 1e-10*max_x)
+        {
+          const double xi = (x + subexpr0)/xi_denominator0;
+          return Point<2>(xi, eta);
+        }
+      else
+        {
+          const double max_y = std::max(std::max(std::abs(y0), std::abs(y1)),
+                                        std::max(std::abs(y2), std::abs(y3)));
+          const double subexpr1 = -eta*y2 + y0*(eta - 1);
+          const double xi_denominator1 = eta*y3 - y1*(eta - 1) + subexpr1;
+          if (std::abs(xi_denominator1) > 1e-10*max_y)
+            {
+              const double xi = (subexpr1 + y)/xi_denominator1;
+              return Point<2>(xi, eta);
+            }
+          else // give up and try Newton iteration
+            {
+              return Point<2>(2, 2);
+            }
+        }
+    }
+
+
+
+    template<int spacedim>
+    Point<3>
+    transform_real_to_unit_cell
+    (const std_cxx11::array<Point<spacedim>, GeometryInfo<3>::vertices_per_cell> &vertices,
+     const Point<spacedim> &p)
+    {
+      Assert(spacedim == 3, ExcInternalError());
+      // To the best of my knowledge there is no exact formula for this.
+      const double x = p(0);
+      const double y = p(1);
+      const double z = p(2);
+
+      const double x0 = vertices[0](0);
+      const double x1 = vertices[1](0);
+      const double x2 = vertices[2](0);
+      const double x3 = vertices[3](0);
+      const double x4 = vertices[4](0);
+      const double x5 = vertices[5](0);
+      const double x6 = vertices[6](0);
+      const double x7 = vertices[7](0);
+
+      const double y0 = vertices[0](1);
+      const double y1 = vertices[1](1);
+      const double y2 = vertices[2](1);
+      const double y3 = vertices[3](1);
+      const double y4 = vertices[4](1);
+      const double y5 = vertices[5](1);
+      const double y6 = vertices[6](1);
+      const double y7 = vertices[7](1);
+
+      const double z0 = vertices[0](2);
+      const double z1 = vertices[1](2);
+      const double z2 = vertices[2](2);
+      const double z3 = vertices[3](2);
+      const double z4 = vertices[4](2);
+      const double z5 = vertices[5](2);
+      const double z6 = vertices[6](2);
+      const double z7 = vertices[7](2);
+
+      // Amazingly, just guessing the middle of the cell does a far (about 5x
+      // faster) better job than trying to approximate the point with an
+      // interpolation scheme.
+      double xi = 0.5;
+      double eta = 0.5;
+      double zeta = 0.5;
+
+      const unsigned int max_iterations = 20;
+
+      for (unsigned int index = 0; index < max_iterations; ++index)
+        {
+          const double x8 = eta*x3;
+          const double x9 = -eta*x2;
+          const double x10 = eta - 1;
```

Comment:
Here and below, is this code automatically generated? If so, can you also add (as a comment) the original program from which it was created?

Otherwise, all of this is so much easier to write and understand if it were stated in terms of the functionality provided by the `Tensor<1,dim>` and `Tensor<2,dim>` classes which allow you to write all of this in terms of vectors and matrices, their inverses, their determinants, etc.